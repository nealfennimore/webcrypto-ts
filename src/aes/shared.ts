/**
 * Shared code for AES
 * @module
 */

import { getKeyUsagePairsByAlg } from "../key_usages.js";
import * as params from "../params.js";
import * as WebCrypto from "../webcrypto.js";

export interface AesGcmCryptoKey extends CryptoKey {
    _aesGcmKeyBrand: any;
}

export interface AesKwCryptoKey extends CryptoKey {
    _aesKwKeyBrand: any;
}

export interface AesCtrCryptoKey extends CryptoKey {
    _aesCtrKeyBrand: any;
}

export interface AesCbcCryptoKey extends CryptoKey {
    _aesCbcKeyBrand: any;
}

export type AesCryptoKeys =
    | AesCbcCryptoKey
    | AesKwCryptoKey
    | AesGcmCryptoKey
    | AesCtrCryptoKey;

export namespace Alg {
    export enum Mode {
        AES_CBC = "AES-CBC",
        AES_CTR = "AES-CTR",
        AES_GCM = "AES-GCM",
        AES_KW = "AES-KW",
    }

    export type Modes = `${Mode}`;
}

export namespace AesShared {
    export async function generateKey<T extends CryptoKey>(
        algorithm: params.EnforcedAesKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<T> {
        return await WebCrypto.generateKey<T, params.EnforcedAesKeyGenParams>(
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function importKey<T extends CryptoKey>(
        format: KeyFormat,
        key: BufferSource | JsonWebKey,
        algorithm: params.EnforcedAesKeyAlgorithms,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<T> {
        return await WebCrypto.importKey<T, params.EnforcedAesKeyAlgorithms>(
            format as any,
            key as any,
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function exportKey<T extends CryptoKey>(
        format: KeyFormat,
        key: T
    ): Promise<JsonWebKey | ArrayBuffer> {
        return await WebCrypto.exportKey(format as any, key);
    }

    export async function encrypt<T extends CryptoKey>(
        algorithm: Exclude<
            params.EnforcedAesParams,
            params.EnforcedAesKwParams
        >,
        key: T,
        data: BufferSource
    ): Promise<ArrayBuffer> {
        return await WebCrypto.encrypt(algorithm, key, data);
    }

    export async function decrypt<T extends CryptoKey>(
        algorithm: Exclude<
            params.EnforcedAesParams,
            params.EnforcedAesKwParams
        >,
        key: T,
        data: BufferSource
    ): Promise<ArrayBuffer> {
        return await WebCrypto.decrypt(algorithm, key, data);
    }

    export async function wrapKey<T extends CryptoKey>(
        format: KeyFormat,
        key: CryptoKey,
        wrappingkey: T,
        wrapAlgorithm: params.EnforcedAesParams
    ): Promise<ArrayBuffer> {
        return await WebCrypto.wrapKey(
            format as any,
            key,
            wrappingkey,
            wrapAlgorithm
        );
    }
    export async function unwrapKey<T extends CryptoKey>(
        format: KeyFormat,
        wrappedKey: BufferSource,
        wrappedKeyAlgorithm: params.EnforcedImportParams,
        unwrappingKey: T,
        unwrappingKeyAlgorithm: params.EnforcedAesParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<CryptoKey> {
        return await WebCrypto.unwrapKey(
            format as any,
            wrappedKey,
            unwrappingKey,
            unwrappingKeyAlgorithm,
            wrappedKeyAlgorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(wrappedKeyAlgorithm.name)
        );
    }
}
