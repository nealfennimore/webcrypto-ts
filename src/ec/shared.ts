import { WebCrypto } from "../crypto";
import { getKeyUsagePairsByAlg } from "../keyUsages";
import * as params from "../params";

export interface EcKey extends CryptoKey {}
export interface EcKeyPair extends CryptoKeyPair {}

export namespace SharedEc {
    export async function generateKey(
        algorithm: params.EnforcedEcKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<EcKeyPair> {
        return await WebCrypto.generateKey<
            EcKeyPair,
            params.EnforcedEcKeyGenParams
        >(
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function importKey(
        format: KeyFormat,
        keyData: BufferSource | JsonWebKey,
        algorithm: params.EnforcedEcKeyImportParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<EcKey> {
        return await WebCrypto.importKey<
            EcKey,
            params.EnforcedEcKeyImportParams
        >(
            format as any,
            keyData as any,
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function exportKey(
        format: KeyFormat,
        keyData: EcKey
    ): Promise<JsonWebKey | ArrayBuffer> {
        return await WebCrypto.exportKey(format as any, keyData);
    }
}
