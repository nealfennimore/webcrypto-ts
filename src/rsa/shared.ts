import { WebCrypto } from "../crypto";
import { getKeyUsagePairsByAlg } from "../keyUsages";
import * as params from "../params";

export interface RsaKey extends CryptoKey {}
export interface RsaKeyPair extends CryptoKeyPair {}

export namespace RsaShared {
    export async function generateKey(
        algorithm: params.EnforcedRsaHashedKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<RsaKey | RsaKeyPair> {
        return await WebCrypto.generateKey<
            RsaKey | RsaKeyPair,
            params.EnforcedRsaHashedKeyGenParams
        >(
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function importKey(
        format: KeyFormat,
        algorithm: params.EnforcedRsaHashedImportParams,
        keyData: BufferSource | JsonWebKey,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<RsaKey> {
        return await WebCrypto.importKey<
            RsaKey,
            params.EnforcedRsaHashedImportParams
        >(
            format as any,
            keyData as any,
            algorithm,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(algorithm.name)
        );
    }

    export async function exportKey(
        format: KeyFormat,
        keyData: RsaKey
    ): Promise<JsonWebKey | ArrayBuffer> {
        return await WebCrypto.exportKey(format as any, keyData);
    }

    export async function sign(
        algorithm:
            | params.EnforcedRsaPssParams
            | params.EnforcedRsassaPkcs1v15Params,
        keyData: RsaKey,
        data: BufferSource
    ): Promise<ArrayBuffer> {
        return await WebCrypto.sign(algorithm, keyData, data);
    }

    export async function verify(
        algorithm:
            | params.EnforcedRsaPssParams
            | params.EnforcedRsassaPkcs1v15Params,
        keyData: RsaKey,
        signature: BufferSource,
        data: BufferSource
    ): Promise<boolean> {
        return await WebCrypto.verify(algorithm, keyData, signature, data);
    }
}
