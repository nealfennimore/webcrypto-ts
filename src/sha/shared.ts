export interface Sha1ArrayBuffer extends ArrayBuffer {
    _sha1ArrayBufferBrand: any;
}
export interface Sha256ArrayBuffer extends ArrayBuffer {
    _sha256ArrayBufferBrand: any;
}
export interface Sha384ArrayBuffer extends ArrayBuffer {
    _sha384ArrayBufferBrand: any;
}
export interface Sha512ArrayBuffer extends ArrayBuffer {
    _sha512ArrayBufferBrand: any;
}

export type ShaArrayBuffers =
    | Sha1ArrayBuffer
    | Sha256ArrayBuffer
    | Sha384ArrayBuffer
    | Sha512ArrayBuffer;

export namespace ShaShared {
    export const hexify = (digest: ShaArrayBuffers) =>
        Array.from(new Uint8Array(digest))
            .map((b) => b.toString(16).padStart(2, "0"))
            .join("");
}
