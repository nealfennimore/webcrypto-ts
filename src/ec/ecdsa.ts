import * as alg from "../alg";
import { WebCrypto } from "../crypto";
import * as params from "../params";
import { EcKey, SharedEc } from "./shared";

export namespace ECDSA {
    export const generateKey = async (
        namedCurve: alg.EC.Curves = alg.EC.Curve.P_521,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ) =>
        await SharedEc.generateKey(
            { name: alg.EC.Variant.ECDSA, namedCurve },
            extractable,
            keyUsages
        );

    export const importKey = async (
        format: KeyFormat,
        namedCurve: alg.EC.Curves,
        keyData: BufferSource | JsonWebKey,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ) =>
        await SharedEc.importKey(
            format,
            { name: alg.EC.Variant.ECDSA, namedCurve },
            keyData,
            extractable,
            keyUsages
        );

    export const exportKey = SharedEc.exportKey;

    export async function sign(
        hash: params.EnforcedEcdsaParams["hash"],
        keyData: EcKey,
        data: BufferSource
    ): Promise<ArrayBuffer> {
        return await WebCrypto.sign(
            {
                name: alg.EC.Variant.ECDSA,
                hash,
            },
            keyData,
            data
        );
    }

    export async function verify(
        hash: params.EnforcedEcdsaParams["hash"],
        keyData: EcKey,
        signature: BufferSource,
        data: BufferSource
    ): Promise<boolean> {
        return await WebCrypto.verify(
            {
                name: alg.EC.Variant.ECDSA,
                hash,
            },
            keyData,
            signature,
            data
        );
    }
}
