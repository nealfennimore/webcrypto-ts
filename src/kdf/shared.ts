/**
 * Shared code for KDF
 * @module
 */
import type { AesCryptoKeys } from "../aes/index.js";
import { AesProxiedCryptoKeys } from "../aes/shared.js";
import { HmacCryptoKey, HmacProxiedCryptoKey } from "../hmac/index.js";
import { DeriveKeyUsagePair, getKeyUsagePairsByAlg } from "../key_usages.js";
import * as params from "../params.js";
import * as proxy from "../proxy.js";
import * as WebCrypto from "../webcrypto.js";

export interface Pbkdf2KeyMaterial extends CryptoKey {
    _pbkdf2KeyMaterialBrand: any;
}
export interface HkdfKeyMaterial extends CryptoKey {
    _kkdfKeyMaterialBrand: any;
}

export namespace Alg {
    export enum Variant {
        PBKDF2 = "PBKDF2",
        HKDF = "HKDF",
    }
    export type Variants = `${Variant}`;
}

export namespace KdfShared {
    export async function generateKeyMaterial<K extends CryptoKey>(
        format: KeyFormat,
        key: BufferSource,
        algorithm: Alg.Variants,
        extractable: boolean = false
    ): Promise<K> {
        return await WebCrypto.importKey(
            format as any,
            key,
            algorithm,
            extractable,
            DeriveKeyUsagePair
        );
    }

    export async function deriveKey(
        algorithm: params.EnforcedPbkdf2Params | params.EnforcedHkdfParams,
        baseKey: Pbkdf2KeyMaterial | HkdfKeyMaterial,
        derivedKeyType:
            | params.EnforcedAesKeyGenParams
            | params.EnforcedHmacKeyGenParams,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<AesCryptoKeys | HmacCryptoKey> {
        return await WebCrypto.deriveKey<
            AesCryptoKeys | HmacCryptoKey,
            params.EnforcedAesKeyGenParams | params.EnforcedHmacKeyGenParams
        >(
            algorithm,
            baseKey,
            derivedKeyType,
            extractable,
            keyUsages ?? getKeyUsagePairsByAlg(derivedKeyType.name)
        );
    }

    export async function deriveBits(
        algorithm: params.EnforcedPbkdf2Params | params.EnforcedHkdfParams,
        baseKey: Pbkdf2KeyMaterial | HkdfKeyMaterial,
        length: number
    ): Promise<ArrayBuffer> {
        return await WebCrypto.deriveBits(algorithm, baseKey, length);
    }
}

export interface HkdfProxiedKeyMaterial
    extends proxy.ProxiedCryptoKey<HkdfKeyMaterial> {
    deriveKey(
        algorithm: Omit<params.EnforcedHkdfParams, "name">,
        derivedKeyType:
            | params.EnforcedAesKeyGenParams
            | params.EnforcedHmacKeyGenParams,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ): Promise<AesProxiedCryptoKeys | HmacProxiedCryptoKey>;

    deriveBits(
        algorithm: Omit<params.EnforcedHkdfParams, "name">,
        length: number
    ): Promise<ArrayBuffer>;
    exportKey: (format: KeyFormat) => Promise<JsonWebKey | ArrayBuffer>;
}

export interface Pbkdf2ProxiedKeyMaterial
    extends proxy.ProxiedCryptoKey<Pbkdf2KeyMaterial> {
    deriveKey(
        algorithm: Omit<params.EnforcedPbkdf2Params, "name" | "iterations">,
        derivedKeyType:
            | params.EnforcedAesKeyGenParams
            | params.EnforcedHmacKeyGenParams,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ): Promise<AesProxiedCryptoKeys | HmacProxiedCryptoKey>;
    deriveBits(
        algorithm: Omit<params.EnforcedPbkdf2Params, "name" | "iterations">,
        length: number
    ): Promise<ArrayBuffer>;
    exportKey: (format: KeyFormat) => Promise<JsonWebKey | ArrayBuffer>;
}
