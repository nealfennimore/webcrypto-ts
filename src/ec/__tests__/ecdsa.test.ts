import { ECDSA } from "../ecdsa";
import { EcKeyPair } from "../shared";

describe("ECDSA", () => {
    let keyPair: EcKeyPair;
    beforeEach(async () => {
        keyPair = await ECDSA.generateKey();
    });
    it("should generate key", async () => {
        expect(keyPair).toMatchSnapshot();
    });
    it("should import and export key", async () => {
        let jwk = await ECDSA.exportKey("jwk", keyPair.publicKey);
        const importedPubKey = await ECDSA.importKey(
            "jwk",
            "P-521",
            jwk,
            true,
            ["verify"]
        );

        expect(await ECDSA.exportKey("jwk", importedPubKey)).toEqual(jwk);

        jwk = await ECDSA.exportKey("jwk", keyPair.privateKey);
        const importedPrivKey = await ECDSA.importKey(
            "jwk",
            "P-521",
            jwk,
            true,
            ["sign"]
        );

        expect(await ECDSA.exportKey("jwk", importedPrivKey)).toEqual(jwk);
    });
    it("should sign and verify", async () => {
        const text = encode("a message");
        const signature = await ECDSA.sign("SHA-512", keyPair.privateKey, text);

        expect(
            await ECDSA.verify("SHA-512", keyPair.publicKey, signature, text)
        ).toBe(true);
    });
});
