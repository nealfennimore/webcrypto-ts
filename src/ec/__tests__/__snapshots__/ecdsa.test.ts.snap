// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ECDSA should generate key 1`] = `
Object {
  "privateKey": CryptoKey {
    Symbol(kKeyObject): PrivateKeyObject {
      Symbol(kKeyType): "private",
    },
    Symbol(kAlgorithm): Object {
      "name": "ECDSA",
      "namedCurve": "P-521",
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "sign",
    ],
  },
  "publicKey": CryptoKey {
    Symbol(kKeyObject): PublicKeyObject {
      Symbol(kKeyType): "public",
    },
    Symbol(kAlgorithm): Object {
      "name": "ECDSA",
      "namedCurve": "P-521",
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "verify",
    ],
  },
}
`;
